### ISSUE TITLE
file 'data' event isn't triggered, 'end' does, but only for files uploaded from an iPhone (?!)

### ISSUE THREAD

**Issue by OpherV** [mscdex/busboy#248]
I'm using busboy as part of the pipeline for our company's [CV submission](https://eko.engineering/letstalkdev) funnel

It works great, except sometimes I'm not getting any file data, this is like 5% of cases.
What I see from the logs of these cases is that the file `end` event is called but not the file `data` event. (When it works I get both).  What's common about all these occurrences is that they all have user agents from an iPhone:

eg:
* `Mozilla/5.0 (iPhone; CPU iPhone OS 13_5 like Mac OS X) AppleWebKit/605.1.15 (KHTML, like Gecko) Mobile/15E148 [FBAN/FBIOS;FBDV/iPhone11,2;FBMD/iPhone;FBSN/iOS;FBSV/13.5;FBSS/3;FBID/phone;FBLC/en_US;FBOP/5]`

* `Mozilla/5.0 (iPhone; CPU iPhone OS 14_6 like Mac OS X) AppleWebKit/605.1.15 (KHTML, like Gecko) Mobile/15E148 [FBAN/FBIOS;FBDV/iPhone12,1;FBMD/iPhone;FBSN/iOS;FBSV/14.6;FBSS/2;FBID/phone;FBLC/he_IL;FBOP/5]`

* `Mozilla/5.0 (iPhone; CPU iPhone OS 14_7 like Mac OS X) AppleWebKit/605.1.15 (KHTML, like Gecko) Version/14.1.2 Mobile/15E148 Safari/604.1`

I myself can't reproduce this which makes debugging really difficult. 
**Any ideas on what could be causing this or what further data can I provide to help resolve this**?

This is my code:
```
const parser = (event) => new Promise((resolve, reject) => {

    let normalizedHeaders = Object.keys(event.headers).reduce((newHeaders, key) => {
        newHeaders[key.toLowerCase()] = event.headers[key];
        return newHeaders;
    }, {});

    const busboy = new Busboy({
        headers: normalizedHeaders
    });

    let result = {
        files: []
    };

    busboy.on('file', (fieldname, file, filename, encoding, mimetype) => {
        console.log('busboy got file', fieldname, filename, encoding, mimetype);

        file.on('data', data => {
            console.log('busboy got file data');

            result.files.push({
                file: data,
                fileName: filename,
                contentType: mimetype
            });
        });

        file.on('end', function() {
            console.log('busboy: File [' + fieldname + '] Finished');
        });
    });

    busboy.on('field', (fieldname, value) => {
        try {
            result[fieldname] = JSON.parse(value);
        } catch (err) {
            result[fieldname] = value;
        }
    });

    busboy.on('error', error => reject(`Parse error: ${error}`));
    busboy.on('finish', () => {
        event.body = result;
        resolve(event);
    });

    busboy.write(event.body, event.isBase64Encoded ? 'base64' : 'binary');
    busboy.end();
});

```

Original full source is here:
https://github.com/EkoLabs/eko-engineering-site/blob/master/lambdas/greenhouse/index.js

Here's a dump of the event headers when this happens (with some redacted personal identifying information):

```
{
    "resource": "/eko-engineering-greenhouse",
    "path": "/eko-engineering-greenhouse",
    "httpMethod": "POST",
    "headers": {
        "accept": "*/*",
        "accept-encoding": "gzip, deflate, br",
        "accept-language": "en-us",
        "content-type": "multipart/form-data; boundary=----WebKitFormBoundarygYCkMtAj0yJZ2d8x",
        "Host": "04g1c14c98.execute-api.us-east-1.amazonaws.com",
        "origin": "https://eko.engineering",
        "referer": "https://eko.engineering/letstalkdev/?utm_source=Facebook&utm_medium=Facebook%20Ad&utm_campaign=Facebook_July2021_v01&fbclid=IwAR3_9eNO0Q3CGCezpbFufuCUY_QrRuGxSz4tV5eNH4EUB9ZEZ-uw3AhZ2oU_aem_AS8jB41wa9usBlOW0wZhCG431ByzQahpyyiys97U_eBGpcKhFMyBBnpGK6_hT6Gi-LkBG6tE7Meof02k8gt39rxLu7WLQUMmzAWaYMI61jORXpQmVtcIWAYdef5Vj1AgKew",
        "User-Agent": "Mozilla/5.0 (iPhone; CPU iPhone OS 13_5 like Mac OS X) AppleWebKit/605.1.15 (KHTML, like Gecko) Mobile/15E148 [FBAN/FBIOS;FBDV/iPhone11,2;FBMD/iPhone;FBSN/iOS;FBSV/13.5;FBSS/3;FBID/phone;FBLC/en_US;FBOP/5]",
        "X-Amzn-Trace-Id": "REDACTED",
        "X-Forwarded-For": "46.19.85.5",
        "X-Forwarded-Port": "443",
        "X-Forwarded-Proto": "https"
    },
    "multiValueHeaders": {
        "accept": [
            "*/*"
        ],
        "accept-encoding": [
            "gzip, deflate, br"
        ],
        "accept-language": [
            "en-us"
        ],
        "content-type": [
            "multipart/form-data; boundary=----WebKitFormBoundarygYCkMtAj0yJZ2d8x"
        ],
        "Host": [
            "04g1c14c98.execute-api.us-east-1.amazonaws.com"
        ],
        "origin": [
            "https://eko.engineering"
        ],
        "referer": [
            "https://eko.engineering/letstalkdev/?utm_source=Facebook&utm_medium=Facebook%20Ad&utm_campaign=Facebook_July2021_v01&fbclid=IwAR3_9eNO0Q3CGCezpbFufuCUY_QrRuGxSz4tV5eNH4EUB9ZEZ-uw3AhZ2oU_aem_AS8jB41wa9usBlOW0wZhCG431ByzQahpyyiys97U_eBGpcKhFMyBBnpGK6_hT6Gi-LkBG6tE7Meof02k8gt39rxLu7WLQUMmzAWaYMI61jORXpQmVtcIWAYdef5Vj1AgKew"
        ],
        "User-Agent": [
            "Mozilla/5.0 (iPhone; CPU iPhone OS 13_5 like Mac OS X) AppleWebKit/605.1.15 (KHTML, like Gecko) Mobile/15E148 [FBAN/FBIOS;FBDV/iPhone11,2;FBMD/iPhone;FBSN/iOS;FBSV/13.5;FBSS/3;FBID/phone;FBLC/en_US;FBOP/5]"
        ],
        "X-Amzn-Trace-Id": [
            "REDACTED"
        ],
        "X-Forwarded-For": [
            "46.19.85.5"
        ],
        "X-Forwarded-Port": [
            "443"
        ],
        "X-Forwarded-Proto": [
            "https"
        ]
    },
    "queryStringParameters": null,
    "multiValueQueryStringParameters": null,
    "pathParameters": null,
    "stageVariables": null,
    "requestContext": {
        "resourceId": "REDACTED",
        "resourcePath": "/eko-engineering-greenhouse",
        "httpMethod": "POST",
        "extendedRequestId": "REDACTED",
        "requestTime": "29/Jul/2021:09:22:40 +0000",
        "path": "/default/eko-engineering-greenhouse",
        "accountId": "REDACTED",
        "protocol": "HTTP/1.1",
        "stage": "default",
        "domainPrefix": "04g1c14c98",
        "requestTimeEpoch": 1627550560833,
        "requestId": "REDACTED",
        "identity": {
            "cognitoIdentityPoolId": null,
            "accountId": null,
            "cognitoIdentityId": null,
            "caller": null,
            "sourceIp": "REDACTED",
            "principalOrgId": null,
            "accessKey": null,
            "cognitoAuthenticationType": null,
            "cognitoAuthenticationProvider": null,
            "userArn": null,
            "userAgent": "Mozilla/5.0 (iPhone; CPU iPhone OS 13_5 like Mac OS X) AppleWebKit/605.1.15 (KHTML, like Gecko) Mobile/15E148 [FBAN/FBIOS;FBDV/iPhone11,2;FBMD/iPhone;FBSN/iOS;FBSV/13.5;FBSS/3;FBID/phone;FBLC/en_US;FBOP/5]",
            "user": null
        },
        "domainName": "04g1c14c98.execute-api.us-east-1.amazonaws.com",
        "apiId": "REDACTED"
    },
    "body": {
        "files": [],
        "utm_medium": "Facebook Ad",
        "utm_source": "Facebook",
        "utm_campaign": "Facebook_July2021_v01",
        "jobId": 5016885002,
        "positionTitle": "Senior Fullstack Developer",
        "name": "REDACTED",
        "email": "REDACTED",
        "phone": "REDACTED",
        "howDidYouHearAboutUs": "Facebook",
        "message": "Self motivated, self learner.\r\nVery friendly.\r\n\r\nOwns a doggo.",
        "site": "Tel Aviv",
        "_date": "29.7.2021, 12:22:40"
    },
    "isBase64Encoded": true
}

```

**Comment by mscdex** [comment]
No idea. Have you verified the contents of `event.body` to see what the individual file part that's failing looks like?

**Comment by OpherV** [comment]
Found it!
So the event.body looked liked this
```
------WebKitFormBoundaryTGhpmEALtUTkeNew
Content-Disposition: form-data; name="utm_medium"

Facebook QA test
------WebKitFormBoundaryTGhpmEALtUTkeNew
Content-Disposition: form-data; name="utm_source"

FacebookTest
------WebKitFormBoundaryTGhpmEALtUTkeNew
Content-Disposition: form-data; name="utm_campaign"

QATest1
------WebKitFormBoundaryTGhpmEALtUTkeNew
Content-Disposition: form-data; name="jobId"

5016885002
------WebKitFormBoundaryTGhpmEALtUTkeNew
Content-Disposition: form-data; name="positionTitle"

Senior Fullstack Developer
------WebKitFormBoundaryTGhpmEALtUTkeNew
Content-Disposition: form-data; name="name"

Qatest
------WebKitFormBoundaryTGhpmEALtUTkeNew
Content-Disposition: form-data; name="email"

intertest100+36@gmail.com
------WebKitFormBoundaryTGhpmEALtUTkeNew
Content-Disposition: form-data; name="phone"

XXXXX
------WebKitFormBoundaryTGhpmEALtUTkeNew
Content-Disposition: form-data; name="howDidYouHearAboutUs"

Test
------WebKitFormBoundaryTGhpmEALtUTkeNew
Content-Disposition: form-data; name="message"

Test
------WebKitFormBoundaryTGhpmEALtUTkeNew
Content-Disposition: form-data; name="site"

Tel Aviv
------WebKitFormBoundaryTGhpmEALtUTkeNew
Content-Disposition: form-data; name="cv"; filename="XXX.pdf"
Content-Type: application/pdf


------WebKitFormBoundaryTGhpmEALtUTkeNew
Content-Disposition: form-data; name="_date"

4.8.2021, 13:41:41
------WebKitFormBoundaryTGhpmEALtUTkeNew--
```


As you can see the body is missing the actual file content.

After a LONG time we figured out it has to do with this stupid safari bug - if you wait more than a minute after selecting a file, then submit the form - the file content will be missing:

https://www.reddit.com/r/apple/comments/bv5vu0/psa_basic_html_file_uploads_are_broken_in_ios/
https://origin-discussions2-us-dr-prz.apple.com/en/thread/252811015
https://discussions.apple.com/thread/250392238

So basically an iOS/Safari client bug, nothing to do with busyboy, interesting nonetheless.
